File: tests/src/test/javascript/entityauth/RsaAuthenticationFactoryTest.js
/**
 * Copyright (c) 2012-2017 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * RSA asymmetric keys entity authentication factory unit tests.
 * 
 * @author Wesley Miaw <wmiaw@netflix.com>
 */
describe("RsaAuthenticationFactory", function() {
    var MslEncoderFormat = require('msl-core/io/MslEncoderFormat.js');
    var EntityAuthenticationScheme = require('msl-core/entityauth/EntityAuthenticationScheme.js');
    var RsaStore = require('msl-core/entityauth/RsaStore.js');
    var RsaAuthenticationFactory = require('msl-core/entityauth/RsaAuthenticationFactory.js');
    var RsaAuthenticationData = require('msl-core/entityauth/RsaAuthenticationData.js');
    var MslEncoderUtils = require('msl-core/io/MslEncoderUtils.js');
    var MslEncodingException = require('msl-core/MslEncodingException.js');
    var MslEntityAuthException = require('msl-core/MslEntityAuthException.js');
    var MslError = require('msl-core/MslError.js');

    var MslTestConstants = require('msl-tests/MslTestConstants.js');
    var MockMslContext = require('msl-tests/util/MockMslContext.js');
    var MockRsaAuthenticationFactory = require('msl-tests/entityauth/MockRsaAuthenticationFactory.js');
    var MslTestUtils = require('msl-tests/util/MslTestUtils.js');
    
    /** MSL encoder format. */
    var ENCODER_FORMAT = MslEncoderFormat.JSON;
    
    /** Key entity identity. */
    var KEY_IDENTITY = "identity";

    /** MSL context. */
    var ctx;
    /** MSL encoder factory. */
    var encoder;
    /** Entity authentication factory. */
    var factory;
    
    var initialized = false;
    beforeEach(function() {
        if (!initialized) {
            runs(function() {
                MockMslContext.create(EntityAuthenticationScheme.RSA, false, {
                    result: function(c) { ctx = c; },
                    error: function(e) { expect(function() { throw e; }).not.toThrow(); }
                });
            });
            waitsFor(function() { return ctx; }, "ctx", MslTestConstants.TIMEOUT_CTX);
            runs(function() {
                encoder = ctx.getMslEncoderFactory();
                var rsaStore = new RsaStore();
                rsaStore.addPublicKey(MockRsaAuthenticationFactory.RSA_PUBKEY_ID, MockRsaAuthenticationFactory.RSA_PUBKEY);
                factory = new RsaAuthenticationFactory(null, rsaStore);
                ctx.addEntityAuthenticationFactory(factory);
                
                initialized = true;
            });
        }
    });
    
    it("createData", function() {
        var data = new RsaAuthenticationData(MockRsaAuthenticationFactory.RSA_ESN, MockRsaAuthenticationFactory.RSA_PUBKEY_ID);
        var entityAuthMo;
        runs(function() {
            data.getAuthData(encoder, ENCODER_FORMAT, {
                result: function(x) { entityAuthMo = x; },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); }
            });
        });
        waitsFor(function() { return entityAuthMo; }, "entityAuthMo", MslTestConstants.TIMEOUT);

        var authdata;
        runs(function() {
            factory.createData(ctx, entityAuthMo, {
                result: function(x) { authdata = x; },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); }
            });
        });
        waitsFor(function() { return authdata; }, "authdata", MslTestConstants.TIMEOUT);

        var dataMo, authdataMo;
        runs(function() {
            expect(authdata).not.toBeNull();
            expect(authdata instanceof RsaAuthenticationData).toBeTruthy();
            
            MslTestUtils.toMslObject(encoder, data, {
                result: function(x) { dataMo = x; },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); }
            });
            MslTestUtils.toMslObject(encoder, authdata, {
                result: function(x) { authdataMo = x; },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); }
            });
        });
        waitsFor(function() { return dataMo && authdataMo; }, "dataMo && authdataMo", MslTestConstants.TIMEOUT);

        runs(function() {
            expect(MslEncoderUtils.equalObjects(dataMo, authdataMo)).toBeTruthy();
        });
    });
    
    it("encode exception", function() {
        var entityAuthMo;
        runs(function() {
	        var data = new RsaAuthenticationData(MockRsaAuthenticationFactory.RSA_ESN, MockRsaAuthenticationFactory.RSA_PUBKEY_ID);
	        data.getAuthData(encoder, ENCODER_FORMAT, {
                result: function(x) { entityAuthMo = x; },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); }
            });
        });
        waitsFor(function() { return entityAuthMo; }, "entityAuthMo", MslTestConstants.TIMEOUT);
        
        var exception;
        runs(function() {
            entityAuthMo.remove(KEY_IDENTITY);
            factory.createData(ctx, entityAuthMo, {
                result: function() {},
                error: function(e) { exception = e; },
            });
        });
        waitsFor(function() { return exception; }, "exception", MslTestConstants.TIMEOUT);
        
        runs(function() {
            var f = function() { throw exception; };
            expect(f).toThrow(new MslEncodingException(MslError.MSL_PARSE_ERROR));
        });
    });
    
    it("crypto context", function() {
        var data = new RsaAuthenticationData(MockRsaAuthenticationFactory.RSA_ESN, MockRsaAuthenticationFactory.RSA_PUBKEY_ID);
        var cryptoContext = factory.getCryptoContext(ctx, data);
        expect(cryptoContext).not.toBeNull();
    });

    it("unknown key ID", function() {
        var f = function() {
	        var data = new RsaAuthenticationData(MockRsaAuthenticationFactory.RSA_ESN, "x");
	        factory.getCryptoContext(ctx, data);
	    };
        expect(f).toThrow(new MslEntityAuthException(MslError.RSA_PUBLICKEY_NOT_FOUND));
    });
    
    it("local crypto context", function() {
        var rsaStore = new RsaStore();
        rsaStore.addPrivateKey(MockRsaAuthenticationFactory.RSA_PUBKEY_ID, MockRsaAuthenticationFactory.RSA_PRIVKEY);
        factory = new RsaAuthenticationFactory(MockRsaAuthenticationFactory.RSA_PUBKEY_ID, rsaStore);
        ctx.addEntityAuthenticationFactory(factory);
        
        var data = new RsaAuthenticationData(MockRsaAuthenticationFactory.RSA_ESN, MockRsaAuthenticationFactory.RSA_PUBKEY_ID);
        var cryptoContext = factory.getCryptoContext(ctx, data);
        
        var plaintext = new Uint8Array(16);
        ctx.getRandom().nextBytes(plaintext);
        cryptoContext.sign(plaintext, encoder, ENCODER_FORMAT, {
            result: function(ciphertext) {},
            error: function(e) { expect(function() { throw e; }).not.toThrow(); }
        });
    });
    
    it("missing private key", function() {
        var f = function() {
            var rsaStore = new RsaStore();
            rsaStore.addPublicKey(MockRsaAuthenticationFactory.RSA_PUBKEY_ID, MockRsaAuthenticationFactory.RSA_PUBKEY);
            var factory = new RsaAuthenticationFactory(MockRsaAuthenticationFactory.RSA_PUBKEY_ID, rsaStore);
            
            var data = new RsaAuthenticationData(MockRsaAuthenticationFactory.RSA_ESN, MockRsaAuthenticationFactory.RSA_PUBKEY_ID);
            factory.getCryptoContext(ctx, data);
        };
        expect(f).toThrow(new MslEntityAuthException(MslError.RSA_PRIVATEKEY_NOT_FOUND));
    });
});

File: tests/src/test/cpp/entityauth/RsaAuthenticationFactoryTest.cpp
/**
 * Copyright (c) 2012-2017 Netflix, Inc.  All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <MslCryptoException.h>
#include <MslEncodingException.h>
#include <MslEntityAuthException.h>
#include <MslError.h>
#include <crypto/ICryptoContext.h>
#include <entityauth/RsaAuthenticationData.h>
#include <entityauth/RsaAuthenticationFactory.h>
#include <io/MslEncoderException.h>
#include <io/MslEncoderFactory.h>
#include <io/MslEncoderFormat.h>
#include <io/MslEncoderUtils.h>
#include <io/MslObject.h>
#include <memory>

#include <entityauth/MockRsaAuthenticationFactory.h>
#include <entityauth/MockRsaStore.h>
#include <util/MockAuthenticationUtils.h>
#include <util/MockMslContext.h>
#include <util/MslTestUtils.h>

using namespace std;
using namespace testing;
using namespace netflix::msl::crypto;
using namespace netflix::msl::io;
using namespace netflix::msl::util;

namespace netflix {
namespace msl {

typedef vector<uint8_t> ByteArray;

namespace entityauth {

namespace {

/** Key entity identity. */
const string KEY_IDENTITY = "identity";

} // namespace anonymous

/**
 * RSA asymmetric keys entity authentication factory unit tests.
 *
 * @author Wesley Miaw <wmiaw@netflix.com>
 */
class RsaAuthenticationFactoryTest : public ::testing::Test
{
public:
	RsaAuthenticationFactoryTest()
		: ctx(make_shared<MockMslContext>(EntityAuthenticationScheme::RSA, false))
		, encoder(ctx->getMslEncoderFactory())
		, format(MslEncoderFormat::JSON)
		, authutils(make_shared<MockAuthenticationUtils>())
	{
		shared_ptr<MockRsaStore> rsaStore = make_shared<MockRsaStore>();
		rsaStore->addPublicKey(MockRsaAuthenticationFactory::RSA_PUBKEY_ID, MockRsaAuthenticationFactory::RSA_PUBKEY);
		factory = make_shared<RsaAuthenticationFactory>(rsaStore, authutils);
		ctx->addEntityAuthenticationFactory(factory);
	}

protected:
	/** MSL context. */
	shared_ptr<MockMslContext> ctx;
	/** MSL encoder factory. */
	shared_ptr<MslEncoderFactory> encoder;
	/** MSL encoder format. */
	const MslEncoderFormat format;
    /** Authentication utilities. */
    shared_ptr<MockAuthenticationUtils> authutils;
    /** Entity authentication factory. */
    shared_ptr<EntityAuthenticationFactory> factory;
};

TEST_F(RsaAuthenticationFactoryTest, createData)
{
	shared_ptr<RsaAuthenticationData> data = make_shared<RsaAuthenticationData>(MockRsaAuthenticationFactory::RSA_ESN, MockRsaAuthenticationFactory::RSA_PUBKEY_ID);
	shared_ptr<MslObject> entityAuthMo = data->getAuthData(encoder, format);

	shared_ptr<EntityAuthenticationData> authdata = factory->createData(ctx, entityAuthMo);
	EXPECT_TRUE(authdata);
	EXPECT_TRUE(instanceof<RsaAuthenticationData>(authdata.get()));

	shared_ptr<MslObject> dataMo = MslTestUtils::toMslObject(encoder, data);
	shared_ptr<MslObject> authdataMo = MslTestUtils::toMslObject(encoder, authdata);
	EXPECT_TRUE(MslEncoderUtils::equalObjects(dataMo, authdataMo));
}

TEST_F(RsaAuthenticationFactoryTest, encodeException)
{
	shared_ptr<RsaAuthenticationData> data = make_shared<RsaAuthenticationData>(MockRsaAuthenticationFactory::RSA_ESN, MockRsaAuthenticationFactory::RSA_PUBKEY_ID);
	shared_ptr<MslObject> entityAuthMo = data->getAuthData(encoder, format);
	EXPECT_FALSE(entityAuthMo->remove(KEY_IDENTITY).isNull());
	try {
		factory->createData(ctx, entityAuthMo);
		ADD_FAILURE() << "Should have thrown";
	} catch (const MslEncodingException& e) {
		EXPECT_EQ(MslError::MSL_PARSE_ERROR, e.getError());
	}
}

TEST_F(RsaAuthenticationFactoryTest, cryptoContext)
{
	shared_ptr<RsaAuthenticationData> data = make_shared<RsaAuthenticationData>(MockRsaAuthenticationFactory::RSA_ESN, MockRsaAuthenticationFactory::RSA_PUBKEY_ID);
	shared_ptr<ICryptoContext> cryptoContext = factory->getCryptoContext(ctx, data);
	EXPECT_TRUE(cryptoContext);
}

TEST_F(RsaAuthenticationFactoryTest, unknownKeyId)
{
	shared_ptr<RsaAuthenticationData> data = make_shared<RsaAuthenticationData>(MockRsaAuthenticationFactory::RSA_ESN, "x");
	try {
		factory->getCryptoContext(ctx, data);
		ADD_FAILURE() << "Should have thrown";
	} catch (const MslEntityAuthException& e) {
		EXPECT_EQ(MslError::RSA_PUBLICKEY_NOT_FOUND, e.getError());
	}
}

TEST_F(RsaAuthenticationFactoryTest, localCryptoContext)
{
	shared_ptr<MockRsaStore> rsaStore = make_shared<MockRsaStore>();
	rsaStore->addPrivateKey(MockRsaAuthenticationFactory::RSA_PUBKEY_ID, MockRsaAuthenticationFactory::RSA_PRIVKEY);
	shared_ptr<EntityAuthenticationFactory> factory = make_shared<RsaAuthenticationFactory>(MockRsaAuthenticationFactory::RSA_PUBKEY_ID, rsaStore, authutils);

	shared_ptr<RsaAuthenticationData> data = make_shared<RsaAuthenticationData>(MockRsaAuthenticationFactory::RSA_ESN, MockRsaAuthenticationFactory::RSA_PUBKEY_ID);
	shared_ptr<ICryptoContext> cryptoContext = factory->getCryptoContext(ctx, data);

	shared_ptr<ByteArray> plaintext = make_shared<ByteArray>(16);
	ctx->getRandom()->nextBytes(*plaintext);
	shared_ptr<ByteArray> signature = cryptoContext->sign(plaintext, encoder, format);
}

TEST_F(RsaAuthenticationFactoryTest, missingPrivateKey)
{
	shared_ptr<MockRsaStore> rsaStore = make_shared<MockRsaStore>();
	rsaStore->addPublicKey(MockRsaAuthenticationFactory::RSA_PUBKEY_ID, MockRsaAuthenticationFactory::RSA_PUBKEY);
	shared_ptr<EntityAuthenticationFactory> factory = make_shared<RsaAuthenticationFactory>(MockRsaAuthenticationFactory::RSA_PUBKEY_ID, rsaStore, authutils);

	shared_ptr<RsaAuthenticationData> data = make_shared<RsaAuthenticationData>(MockRsaAuthenticationFactory::RSA_ESN, MockRsaAuthenticationFactory::RSA_PUBKEY_ID);
	try {
		factory->getCryptoContext(ctx, data);
		ADD_FAILURE() << "Should have thrown";
	} catch (const MslEntityAuthException& e) {
		EXPECT_EQ(MslError::RSA_PRIVATEKEY_NOT_FOUND, e.getError());
	}
}

TEST_F(RsaAuthenticationFactoryTest, revoked)
{
	authutils->revokeEntity(MockRsaAuthenticationFactory::RSA_ESN);
	shared_ptr<RsaAuthenticationData> data = make_shared<RsaAuthenticationData>(MockRsaAuthenticationFactory::RSA_ESN, "x");
	try {
		factory->getCryptoContext(ctx, data);
		ADD_FAILURE() << "Should have thrown";
	} catch (const MslEntityAuthException& e) {
		EXPECT_EQ(MslError::ENTITY_REVOKED, e.getError());
	}
}

}}} // namespace netflix::msl::entityauth
